### PR TITLE
[Feature]: Vector Index | Phase 4 (Handle Nulls, Alter Table etc)

### PR DESCRIPTION
**Pull request by arjunsk**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [X] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue # https://github.com/matrixorigin/matrixone/issues/10634

## What this PR does / why we need it:
- Don't delete old version number entries. (removing unused code)
- Fix "insert into select *" issue by introducing `Window Node`.
- Fix alter table add column issue. (reindex after table creation and insert)
- Create Index with `no rows` will succeed with NULL centroids. Create index with `totalNonNullCnt < k` will succeed with NULL centroids.
- Handle Nulls in vector index. NULL vector is assigned to any centroid.

### CODE REVIEW

**Comment by matrix-meow** [comment]
@arjunsk Thanks for your contributions!

**Here are review comments for file pkg/catalog/secondary_index_utils.go:**
Review:

Title: [Feature]: Vector Index | Phase 4 (Handle Nulls, Alter Table etc)

Body:
- The PR is a feature implementation.
- It fixes issues related to version rotation, "insert into select *" statement, and alter table add column.
- It allows creating an index with no rows or with totalNonNullCnt < k.
- It handles nulls in vector index.

Changes in pkg/catalog/secondary_index_utils.go:
- Line 76: The code changes the value of `sampleCnt` from `10000` to `10_000`.

Overall, the title and body of the pull request provide a clear overview of the changes made and the issues addressed. The changes in the code seem to be focused on fixing specific issues related to the vector index feature.

However, there are a few areas that need attention:

1. In the body of the pull request, it would be helpful to provide more context on the specific issues that were fixed. For example, mentioning the specific problems with version rotation, "insert into select *" statement, and alter table add column would provide more clarity.

2. The code change in `secondary_index_utils.go` seems to be a minor adjustment to the `CalcSampleCount` function. However, the reasoning behind this change is not explained in the pull request. It would be beneficial to include a brief explanation of why the value of `sampleCnt` is being changed from `10000` to `10_000`. This will help reviewers understand the intention behind the change and ensure that it is correct.

3. Security: There are no apparent security issues in the provided code changes.

To optimize the changes made in the pull request, consider the following suggestions:

1. Provide more detailed explanations in the pull request body for each issue fixed. This will help reviewers understand the context and impact of the changes.

2. Consider adding comments in the code to explain the purpose and reasoning behind the changes. This will make the code more maintainable and easier to understand for future developers.

3. Consider adding test cases to cover the fixed issues. This will ensure that the changes are thoroughly tested and prevent regressions in the future.

Overall, the pull request seems to be addressing specific issues related to the vector index feature. By providing more detailed explanations and adding test cases, the quality and maintainability of the codebase can be improved.

**Here are review comments for file pkg/sql/compile/alter.go:**
Review:

Title: [Feature]: Vector Index | Phase 4 (Handle Nulls, Alter Table etc)

Body:
- The PR is a feature implementation.
- It fixes issues related to version rotation, "insert into select *" statement, and alter table add column.
- It allows creating an index with no rows or with totalNonNullCnt < k.
- It handles nulls in vector index.

Changes in `pkg/sql/compile/alter.go`:
- Added code to invoke reindex for the new table if it contains an ivf index.
- The code creates a map of multi-table indexes and their types.
- It then iterates over the indexes and handles the vector ivf flat index.
- Finally, it deletes the temporary replica table.

Feedback:
- The title and body of the pull request are clear and provide a good overview of the changes.
- The changes in `pkg/sql/compile/alter.go` seem to address the issues mentioned in the body.
- However, there are a few areas that need attention:

1. Code duplication: The code for handling vector ivf flat index is repeated for each index algorithm type. This can be refactored to avoid duplication and improve maintainability. Consider extracting the common code into a separate function and calling it for each index algorithm type.

2. Error handling: The error handling in the code is inconsistent. In some cases, errors are returned immediately, while in others, errors are accumulated and returned at the end. It would be better to have consistent error handling throughout the code. Consider returning errors immediately when encountered and using a single error variable to accumulate errors if needed.

3. Security: The code does not appear to introduce any security issues.

Overall, the changes in the pull request seem to address the mentioned issues and add the desired features. However, the code can be optimized and improved for better maintainability and error handling.

**Here are review comments for file pkg/sql/compile/ddl.go:**
Title: [Feature]: Vector Index | Phase 4 (Handle Nulls, Alter Table etc)

Body:
- This pull request addresses several issues related to vector index functionality.
- It removes version rotation and ensures that old version number entries are not deleted.
- It fixes the "insert into select *" issue.
- It fixes the alter table add column issue.
- It allows the creation of an index with no rows or with totalNonNullCnt < k.
- It handles null values in vector index.

Changes in pkg/sql/compile/ddl.go:
- Lines 1220-1222: The condition `if totalCnt == 0` is removed, which means that the function will continue even if the total count is 0. This may introduce unexpected behavior or errors if the count is expected to be non-zero. It is recommended to add a comment explaining why this condition is removed and ensure that the code handles the case when the count is 0 correctly.
- Lines 1236-1239: The code that deletes old entries in the "centroids" and "entries" tables for the current version is removed. It is unclear why this code is removed and what impact it may have on the functionality. It is recommended to add a comment explaining the reason for removing this code and ensure that the removal does not introduce any issues or security vulnerabilities.
- Lines 1244-1247: The comment says "populate centroids table", but the code actually populates the entries table. This comment should be updated to reflect the correct operation being performed.
- Lines 1252-1255: The comment says "populate entries table", but the code actually populates the centroids table. This comment should be updated to reflect the correct operation being performed.

Optimization suggestions:
- The code could benefit from more descriptive variable and function names to improve readability and maintainability.
- It would be helpful to add comments explaining the purpose and functionality of each section of code.
- Consider refactoring the code to separate the different operations into separate functions or methods to improve modularity and readability.
- Consider adding error handling and logging to provide better feedback and debugging information in case of failures or errors.

**Here are review comments for file pkg/sql/compile/ddl_index_algo.go:**
Review:

Title: [Feature]: Vector Index | Phase 4 (Handle Nulls, Alter Table etc)

Body: The pull request is a feature request that aims to handle null values in vector indexes. It also includes fixes for issues related to inserting into select statements and adding columns to tables. The pull request is associated with issue #10634.

Changes in pkg/sql/compile/ddl_index_algo.go:
- In the `handleIndexAndPKColCount` function, the code has been modified to remove the check for the equality of `totalCnt` and `pkeyCnt`. This check was previously used to ensure that the vector column does not contain nulls. However, this check has been removed without any explanation or replacement logic. This change should be addressed by either reintroducing the null check or providing a clear explanation for why it is no longer necessary.
- The `handleIvfIndexCentroidsTable` function has been modified to add an initialization step for the centroids table if the total count is 0 or less than the required number of centroids. The initialization inserts a default centroid into the table. This change seems reasonable and can help ensure that the centroids table is always populated with the required number of centroids.
- The `handleIvfIndexEntriesTable` function has been modified to remove the unnecessary `nonNullOriginalTableRowsSql` query. This query selects non-null rows from the original table, but it is not used in the subsequent SQL statement. Removing this query improves the efficiency of the code.

Overall, the changes in this pull request seem to address the issues mentioned in the body. However, there is a missing null check in the `handleIndexAndPKColCount` function that needs to be addressed. Additionally, the removal of the `nonNullOriginalTableRowsSql` query in the `handleIvfIndexEntriesTable` function is a good optimization.

**Here are review comments for file pkg/sql/plan/build_dml_util.go:**
Review:

Title: [Feature]: Vector Index | Phase 4 (Handle Nulls, Alter Table etc)

Body:
- The PR is a feature implementation.
- It fixes issues related to version rotation, "insert into select *" behavior, and alter table add column.
- It allows creating an index with no rows or with a totalNonNullCnt less than k.
- It handles nulls in vector index.

Changes in `pkg/sql/plan/build_dml_util.go`:
- Lines 2458-2461: Adds projections for centroids.version, centroids.centroid_id, tbl.pk, centroids.centroid, and tbl.embedding.
- Lines 2464-2469: Creates a cross join node for tbl and centroids.
- Lines 2473-2479: Adds partitioning by tbl.pk.
- Lines 2482-2484: Adds a window operation.
- Lines 2487-2489: Filters records where row_number() = 1.
- Lines 2493-2500: Adds a final projection for centroids.version, centroids.centroid_id, tbl.pk, and cp_col.

Suggestions:
- The title and body of the pull request are clear and provide a good overview of the changes made.
- The changes in `build_dml_util.go` seem to be related to the feature being implemented, but it's difficult to fully understand the context without more information about the codebase and the specific requirements of the feature.
- It would be helpful to have more comments or explanations within the code to provide additional context and make it easier to understand the purpose of each change.
- It's important to ensure that the changes made in this pull request do not introduce any security vulnerabilities. It would be beneficial to conduct a security review of the code changes to identify any potential issues and address them accordingly.
- Additionally, it would be useful to optimize the code changes by considering any potential performance improvements or simplifications that can be made. This could involve analyzing the existing code and identifying any redundant or inefficient operations that can be optimized.

**Here are review comments for file pkg/sql/plan/build_vector_index_util.go:**
Title: [Feature]: Vector Index | Phase 4 (Handle Nulls, Alter Table etc)

Body: This pull request is a feature implementation for vector index. It addresses several issues including removing version rotation, fixing "insert into select *" issue, fixing alter table add column issue, and handling nulls in vector index.

Changes in `pkg/sql/plan/build_vector_index_util.go`:
- The function `makeSortByL2DistAndLimit1AndProject4` has been renamed to `makeFinalProjectWithCPAndOptionalRowId`.
- The function `makeFinalProjectWithCPAndOptionalRowId` now includes the projection of `entries.row_id` if it is an update operation.
- The function `makeFinalProjectWithCPAndOptionalRowId` now returns the ID of the final project node.
- The function `partitionByWindowAndFilterByRowNum` has been added.
- The function `partitionByWindowAndFilterByRowNum` partitions the data by `tbl.pk`, performs a window operation with the `row_number` function, and filters the records where `row_number() = 1`.
- The function `partitionByWindowAndFilterByRowNum` returns the ID of the filter node.

Overall, the changes seem to be focused on improving the functionality and performance of the vector index feature. However, there are a few areas that need attention:

1. In the `makeFinalProjectWithCPAndOptionalRowId` function, the comment for the project list is outdated. It still mentions `3: centroids.centroid` and `4: tbl.embedding`, but these projections are no longer present in the code. The comment should be updated to reflect the current project list.

2. In the `makeFinalProjectWithCPAndOptionalRowId` function, the `l2_distance` expression is bound using the `BindFuncExprImplByPlanExpr` function. It would be more efficient to pre-compile this expression outside of the loop to avoid unnecessary recompilation.

3. In the `partitionByWindowAndFilterByRowNum` function, the comment for the partition by sort key ID node mentions `tbl.pk`, but the actual expression used for partitioning is `projections[2]`. The comment should be updated to match the code.

4. In the `partitionByWindowAndFilterByRowNum` function, the `rowNumberCol` expression is created manually instead of using the `MakePlan2ColRefExpr` function. It would be cleaner and more consistent to use the function for creating the column reference expression.

5. In the `partitionByWindowAndFilterByRowNum` function, the filter node is created with the `FilterList` field set to a single expression. It would be more flexible to use the `FilterExpr` field instead, which allows for more complex filter conditions if needed in the future.

6. In the `partitionByWindowAndFilterByRowNum` function, the filter node only includes the projections `projections[0]`, `projections[1]`, and `projections[2]`. It's unclear why the `rowNumberCol` projection is not included in the final project list. If it's not needed, the comment should be updated to explain why.

7. It's unclear why the `partitionByWindowAndFilterByRowNum` function returns `-1` in the error cases. It would be more informative to return a specific error value or provide a more detailed error message.

To optimize the changes made in the pull request, the following suggestions can be considered:

1. Pre-compile the `l2_distance` expression in the `makeFinalProjectWithCPAndOptionalRowId` function to avoid unnecessary recompilation.

2. Use the `MakePlan2ColRefExpr` function to create the `rowNumberCol` expression in the `partitionByWindowAndFilterByRowNum` function for consistency and readability.

3. Use the `FilterExpr` field instead of the `FilterList` field in the filter node of the `partitionByWindowAndFilterByRowNum` function to allow for more complex filter conditions if needed in the future.

4. Provide more informative error messages or specific error values instead of returning `-1` in the error cases of the `partitionByWindowAndFilterByRowNum` function.

Overall, the changes in the pull request seem to address the issues mentioned in the body and improve the functionality of the vector index feature. However, the areas mentioned above need attention and the suggested optimizations can further improve the code.

**Here are review comments for file test/distributed/cases/array/array_index.result:**
Title: [Feature]: Vector Index | Phase 4 (Handle Nulls, Alter Table etc)

Body: This pull request is a feature implementation that focuses on handling null values in vector indexes. It also includes fixes for issues related to inserting into a table using the "select *" syntax and adding columns to a table. Additionally, it allows for the creation of indexes with no rows or with a total non-null count less than a specified value.

Changes in the file `test/distributed/cases/array/array_index.result`:
- Line 152: The value being inserted into the `tbl` table has been changed from `[100,44,50]` to `[100,44,50]` with a comment indicating that it is a duplicate.
- Line 160: The index `idx8` has been successfully created using the `IVFFLAT` algorithm with the specified options.
- Line 162: The output of the `show index from tbl` command now includes the newly created index `idx8`.
- Line 164: The output of the `show create table tbl` command now includes the definition of the `idx8` index.
- Line 166: The output of the `select name, type, column_name, algo, algo_table_type, algo_params from mo_catalog.mo_indexes where name="idx8"` command now includes the details of the `idx8` index.
- Line 174: The `tbl` table is dropped.
- Line 175: The `tbl` table is created with a new column `id2` of type `VARCHAR(20)`.
- Line 176: The `id2` column is updated with the values from the `id` column.
- Line 178: The `tbl1` table is dropped.
- Line 179: The `tbl1` table is created with columns `id` and `data` of types `int` and `vecf32(3)` respectively.
- Lines 180-187: Rows are inserted into the `tbl1` table.
- Line 188: The `idx19` index is created on the `tbl1` table using the `IVFFLAT` algorithm with the specified options.
- Lines 189-198: Rows are inserted into the `tbl2` table by selecting from the `tbl1` table.
- Line 199: The `tbl1` table is dropped.
- Line 200: The `tbl1` table is created again with columns `id` and `data` of types `int` and `vecf32(3)` respectively.
- Line 201: The `idx19` index is created on the `tbl1` table again using the `IVFFLAT` algorithm with the specified options.
- Lines 202-209: Rows are inserted into the `tbl1` table.
- Line 210: The `tbl` table is dropped.
- Line 211: The `tbl` table is created with columns `id` and `data` of types `int` and `vecf32(3)` respectively.
- Lines 212-216: Rows are inserted into the `tbl` table.
- Line 217: The `idx20` index is created on the `tbl` table using the `IVFFLAT` algorithm with the specified options.
- Lines 218-219: Rows are inserted into the `tbl` table.

Overall, the changes in this pull request seem to address the issues mentioned in the body of the pull request. However, there are a few areas that could be improved:

1. In the body of the pull request, it would be helpful to provide more context on why handling null values in vector indexes is important and how it improves the functionality of the system.

2. The changes in the file `test/distributed/cases/array/array_index.result` seem to be related to testing the new features and fixes. It would be beneficial to have additional test cases that cover different scenarios and edge cases to ensure the robustness of the implementation.

3. It is unclear why the `tbl` table is dropped and created multiple times in the test cases. It would be more efficient to reuse the existing table instead of recreating it.

4. The comments in the test cases could be improved to provide more clarity on the purpose of each test case and the expected outcome.

5. It would be helpful to include some additional documentation or comments in the code to explain the changes made and the reasoning behind them. This would make it easier for future developers to understand and maintain the code.

In terms of security, there are no apparent security issues in the changes made in this pull request. However, it is always a good practice to review the code for potential security vulnerabilities, such as SQL injection or insecure data handling, before merging it into the codebase.

**Here are review comments for file test/distributed/cases/array/array_index.sql:**
Title: [Feature]: Vector Index | Phase 4 (Handle Nulls, Alter Table etc)

Body: 
- The PR removes version rotation and ensures that old version number entries are not deleted.
- It fixes the "insert into select *" issue.
- It fixes the alter table add column issue.
- It allows creating an index with no rows and with totalNonNullCnt < k.
- It handles Nulls in vector index.

Changes for file test/distributed/cases/array/array_index.sql:
- Line 89: The test case for duplicate rows has been modified.
- Lines 91-93: The test case for adding null values to the table has been removed.
- Lines 95-96: The test case for creating an index with no rows has been added.
- Lines 98-109: The test case for handling null embeddings has been added.

Overall, the changes in this pull request address various issues related to vector indexing, including handling null values, fixing insert and alter table issues, and improving index creation with no rows. The changes in the test cases ensure that these functionalities are properly tested.

However, there are a few areas that need attention:

1. Security: The PR does not mention any security-related changes or improvements. It is important to ensure that the changes made do not introduce any security vulnerabilities. It is recommended to conduct a thorough security review of the code changes and address any potential security issues.

2. Code Optimization: The changes in the test cases could be optimized to improve readability and maintainability. For example, instead of dropping and creating the same table multiple times, the test cases could reuse the existing table and perform the necessary modifications.

3. Documentation: The PR does not mention any updates to the documentation. It is important to ensure that any changes made are properly documented to help developers understand the functionality and usage of the vector index feature.

To address these issues:
1. Conduct a security review of the code changes to identify and address any potential security vulnerabilities.
2. Optimize the test cases to reuse the existing table and perform necessary modifications instead of dropping and creating the table multiple times.
3. Update the documentation to reflect the changes made in the pull request and provide clear instructions on how to use the vector index feature.

By addressing these issues, the pull request can be further improved in terms of security, code optimization, and documentation, ensuring a higher quality codebase.

**Review by ouyuanning — APPROVED** [review]

**Review by iamlinjunhong — APPROVED** [review]

**Review by aressu1985 — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Pull request merged**
